rs/sns/governance/src/governance.rs

```diff
@@ -4623,8 +4634,17 @@ impl Governance {
 
         self.process_proposals();
 
-        if self.should_check_upgrade_status() {
-            self.check_upgrade_status().await;
+        if self.acquire_upgrade_periodic_task_lock() {
+            if self.should_check_upgrade_status() {
+                self.check_upgrade_status().await;
+            }
+
+            if self.should_refresh_cached_upgrade_steps() {
+                self.temporarily_lock_refresh_cached_upgrade_steps();
```

Comment:
A comment to explain why we need both lock to guard this refresh task could help.

Reply:
Done!